vowpalwabbit/core/src/reductions/cb/cb_adf.cc

```diff
@@ -337,37 +340,50 @@ void save_load(VW::reductions::cb_adf& c, VW::io_buf& model_file, bool read, boo
   }
 
   std::stringstream msg;
-  msg << "event_sum " << c.get_gen_cs().event_sum << "\n";
-  VW::details::bin_text_read_write_fixed(
-      model_file, (char*)&c.get_gen_cs().event_sum, sizeof(c.get_gen_cs().event_sum), read, msg, text);
+  msg << "event_sum " << c.get_gen_cs().per_model_state[0].event_sum << "\n";
+  VW::details::bin_text_read_write_fixed(model_file, (char*)&c.get_gen_cs().per_model_state[0].event_sum,
+      sizeof(c.get_gen_cs().per_model_state[0].event_sum), read, msg, text);
 
-  msg << "action_sum " << c.get_gen_cs().action_sum << "\n";
-  VW::details::bin_text_read_write_fixed(
-      model_file, (char*)&c.get_gen_cs().action_sum, sizeof(c.get_gen_cs().action_sum), read, msg, text);
+  msg << "action_sum " << c.get_gen_cs().per_model_state[0].action_sum << "\n";
+  VW::details::bin_text_read_write_fixed(model_file, (char*)&c.get_gen_cs().per_model_state[0].action_sum,
+      sizeof(c.get_gen_cs().per_model_state[0].action_sum), read, msg, text);
 }
 
 void cb_adf_merge(const std::vector<float>& /* per_model_weights */,
     const std::vector<const VW::reductions::cb_adf*>& sources, VW::reductions::cb_adf& output_data)
 {
   for (const auto* source : sources)
   {
-    output_data.get_gen_cs().event_sum += source->get_gen_cs().event_sum;
-    output_data.get_gen_cs().action_sum += source->get_gen_cs().action_sum;
+    for (size_t i = 0; i < output_data.get_gen_cs().per_model_state.size(); i++)
+    {
+      output_data.get_gen_cs().per_model_state[i].event_sum += source->get_gen_cs().per_model_state[i].event_sum;
+      output_data.get_gen_cs().per_model_state[i].action_sum += source->get_gen_cs().per_model_state[i].action_sum;
+    }
   }
 }
 
 void cb_adf_add(
     const VW::reductions::cb_adf& data1, const VW::reductions::cb_adf& data2, VW::reductions::cb_adf& data_out)
 {
-  data_out.get_gen_cs().event_sum = data1.get_gen_cs().event_sum + data2.get_gen_cs().event_sum;
-  data_out.get_gen_cs().action_sum = data1.get_gen_cs().action_sum + data2.get_gen_cs().action_sum;
+  for (size_t i = 0; i < data1.get_gen_cs().per_model_state.size(); i++)
+  {
+    data_out.get_gen_cs().per_model_state[i].event_sum =
+        data1.get_gen_cs().per_model_state[i].event_sum + data2.get_gen_cs().per_model_state[i].event_sum;
```

Comment:
this could become c.get_gen_cs(ft_offset=i).event_sum or not sure

Reply:
it would look more similar to the left side of the diff

Reply:
move to seperate pr